packages/vite/src/node/config.ts

```diff
@@ -1780,6 +1789,70 @@ async function nativeImportConfigFile(resolvedPath: string) {
   }
 }
 
+// rather than a singleton, it would be better to unregister the loader
+// so that it doesn't incur the overhead when not needed
+// but unregistering a loader is not supported
+let depsTracker: DepsTracker | undefined
+
+type DepsTracker = {
+  collect: <T>(
+    cb: () => Promise<T>,
+  ) => Promise<{ result: T; dependencies: string[] }>
+}
+
+// This only tracks ESM dependencies that are statically imported (not dynamic imports)
+async function createDepsTracker(): Promise<DepsTracker | undefined> {
+  const { Module } = await import('node:module')
+  // register only exists in Node.js 18.19.0+, 20.6.0+
+  if (!Module.register) {
+    return
+  }
+
+  // we want to register this loader as the last one
+  // this is ensured for the first config load,
+  // but for the subsequent config loads, a different loader may be registered.
+  // we hope that that doesn't happen
+  const { port1, port2 } = new MessageChannel()
+  Module.register('#deps-tracker', {
+    parentURL: import.meta.url,
+    data: { port: port2 },
+    transferList: [port2],
+  })
+  port1.unref()
+
+  let collecting = false
+  return {
+    async collect(cb) {
+      if (collecting) throw new Error('already collecting')
```

Comment:
Actually, it would probably error at this line if two config loads runs concurrently. Propagating `?t=\d&config=\s` should probably fix it.